### PR TITLE
Fix completion after application with implicit arguments

### PR DESCRIPTION
**Pull request by retronym**
`List(1, 2, 3).map(f).<ctrl-space>` now works; before
the tree had the type `(bf: CanBuildFrom[...]):...` and
did not contribute completions from the result type.

This commit checks if the tree has an implicit method
type, and typechecks it as a qualifier. That is enough
to get to `adaptToImplicitMethod` in the type checker,
infer the implicit arguments, and compute the final result
type accordingly.

### CODE REVIEW

**Comment by retronym** [comment]
Review by @dragos, /cc @gkossakowski

**Comment by retronym** [comment]
@dragos: here is the test fixture for scope completion: https://github.com/retronym/scala/compare/scala:2.10.x...ticket%2Fcompletion2?expand=1

**Comment by paulp** [comment]
Nice. Wouldn't it be lovely for this to work in the repl too. (Wouldn't it be lovelier if that were an obvious/unavoidable consequence of making it work in another place.)

**Comment by som-snytt** [comment]
I have previously thought that repl does too much work and should issue interactive commands instead; that's especially clean with the UI on one thread and the compiler on another; I was just looking at completion in repl, but stopped short of looking at what it's doing; I'm still working on... imports.

**Comment by dotta** [comment]
@retronym A Scala IDE user recently reported a suspiciously similar issue https://scala-ide-portfolio.assembla.com/spaces/scala-ide/tickets/1001944

Could you check that the following works as expected with this PR

```
val tupleCollection = List((1, 2, 3))
(for (current <- tupleCollection) yield (current._1, current._2)).toStrea /*!*/ <-- ask completion here, expect `toStream` to be returned
```

**Comment by retronym** [comment]
Confirmed that it works now. I don't think we need to test this variation as it is identical to the type checker:

```
ticket/completion /code/scala2 git diff
diff --git a/test/files/presentation/completion-implicit-chained/src/Completions.scala b/test/files/
index 67922df..692fd78 100644
--- a/test/files/presentation/completion-implicit-chained/src/Completions.scala
+++ b/test/files/presentation/completion-implicit-chained/src/Completions.scala
@@ -8,5 +8,7 @@ object O {
 }

 class Foo {
-  O.map(x => x)./*!*/ // we want the presentation compiler to apply the implicit argument list.
+  // O.map(x => x)./*!*/ // we want the presentation compiler to apply the implicit argument list.
+  val tupleCollection = List((1, 2, 3))
+ (for (current <- tupleCollection) yield (current._1, current._2))./*!*/
 }
ticket/completion /code/scala2 pt --show-log test/files/presentation/completion-implicit-chained | grep toStream
-[accessible:  true] `method toStream=> scala.collection.immutable.Stream[(Int, Int)]`
[accessible:  true] `method toStream=> scala.collection.immutable.Stream[(Int, Int)]`
```

File: src/compiler/scala/tools/nsc/interactive/Global.scala [hunk 1]
```diff
@@ -999,7 +999,13 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
 
     val context = doLocateContext(pos)
 
-    if (tree.tpe == null)
+    val shouldTypeQualifier = tree.tpe match {
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/interactive/Global.scala:1002]: This seems generally useful. When would it make sense to _not_ type the qualifier if it's `MethodType` and the method is implicit? I wonder if there are other places where this test should be done.

`askType` has a similar problem (stopping type-checking to early) for overloaded methods. See [SI-6548](https://issues.scala-lang.org/browse/SI-7548)

**Comment by dragos** [comment]
LGTM, thanks!

File: src/compiler/scala/tools/nsc/interactive/Global.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/interactive/Global.scala:1002]: Maybe this patch isn't the only way to skin this cat. I guess that `typedTreeAt` finds the inner apply in `ApplyImplicitArgs(Apply("map", List("x => x"))(List("canBuildFrom"))`, and we compensate here by typechecking that against a wildcard expected type, which wraps another `ApplyImplicitArgs` node around this.

An alternative fix would be to change `typedTreeAt` to pick a directly enclosing `ApplyImplicitArgs` node, if one exists.

**Pull request merged**